---
source: src/type_checker/mod.rs
assertion_line: 517
expression: asts
input_file: src/type_checker/tests/inputs/call.lt
---
- - "// basic\nfn plusOne(x: int) -> int {\n    x + 1\n}\nfn main() {\n    plusOne(1)\n}\n"
  - Ok:
      nodes:
        - Stmt:
            Fn:
              proto:
                name: plusOne
                args:
                  - - x
                    - Int32
                ret_ty: Int32
              body:
                Block:
                  list:
                    - Expr:
                        BinOp:
                          sym: Add
                          lhs:
                            Expr:
                              Ident:
                                name: x
                                ty: Int32
                          rhs:
                            Expr:
                              Lit:
                                value:
                                  Int32: 1
                                ty: Int32
                          ty: Int32
                  ty: Int32
        - Stmt:
            Fn:
              proto:
                name: main
                args: []
                ret_ty: Int32
              body:
                Block:
                  list:
                    - Expr:
                        Call:
                          name: plusOne
                          args:
                            - Expr:
                                Lit:
                                  value:
                                    Int32: 1
                                  ty: Int32
                          ty: Int32
                  ty: Int32
- - "\n// undefined function\nfn plusOne(x: int) -> int {\n    x + 1\n}\nfn main() {\n    plusThree(1)\n}\n"
  - Err: "Call to undefined function: `plusThree`"
- - "\n// mismatch arg type\nfn plusOne(x: int) -> int {\n    x + 1\n}\nfn main() {\n    let x: float = 1.0\n    plusOne(x)\n}\n"
  - Err: "Type mismatch in arg 1 of call to plusOne: int32 != float"
- - "\n// mismatch arg num\nfn plusOne(x: int) -> int {\n    x + 1\n}\nfn main() {\n    plusOne(1, 2)\n}\n"
  - Err: Call to plusOne takes 1 args and 2 were given

